File: benchmark/matrix_multiplication/matmul.c
#include <stdio.h>
#include <stdlib.h>

int main() {
    int size = 300;
    int total_elements = size * size;
    
    // Using long long to match ADGLANG's 64-bit int
    long long* a = (long long*)malloc(total_elements * sizeof(long long));
    long long* b = (long long*)malloc(total_elements * sizeof(long long));
    long long* c = (long long*)malloc(total_elements * sizeof(long long));
    
    // Initialize
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            int idx = i * size + j;
            a[idx] = i + j;
            b[idx] = i - j;
        }
    }
    
    // Multiply
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            long long sum = 0;
            for (int k = 0; k < size; k++) {
                sum += a[i * size + k] * b[k * size + j];
            }
            c[i * size + j] = sum;
        }
    }
    
    // Verify
    long long result_sum = 0;
    for (int i = 0; i < total_elements; i++) {
        result_sum += c[i];
    }
    
    printf("Matrix %dx%d sum: %lld\n", size, size, result_sum);
    
    free(a);
    free(b);
    free(c);
    
    return 0;
}

File: benchmark/matrix_multiplication/matmul.go
package main

import "fmt"

func main() {
	size := 300
	totalElements := size * size

	a := make([]int64, totalElements)
	b := make([]int64, totalElements)
	c := make([]int64, totalElements)

	// Initialize
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			idx := i*size + j
			a[idx] = int64(i + j)
			b[idx] = int64(i - j)
		}
	}

	// Multiply
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			var sum int64 = 0
			for k := 0; k < size; k++ {
				sum += a[i*size+k] * b[k*size+j]
			}
			c[i*size+j] = sum
		}
	}

	// Verify
	var resultSum int64 = 0
	for i := 0; i < totalElements; i++ {
		resultSum += c[i]
	}

	fmt.Printf("Matrix %dx%d sum: %d\n", size, size, resultSum)
}
